bin/spice/pkg/spec/spicepod.go

```diff
@@ -16,17 +16,73 @@ limitations under the License.
 
 package spec
 
+import (
+	"gopkg.in/yaml.v3"
+)
+
 type SpicepodSpec struct {
-	Version      string            `json:"version,omitempty" csv:"version" yaml:"version,omitempty"`
-	Kind         string            `json:"kind,omitempty" csv:"kind" yaml:"kind,omitempty"`
-	Name         string            `json:"name,omitempty" csv:"name" yaml:"name,omitempty"`
-	Params       map[string]string `json:"params,omitempty" yaml:"params,omitempty" mapstructure:"params,omitempty"`
-	Metadata     map[string]string `json:"metadata,omitempty" csv:"metadata" yaml:"metadata,omitempty"`
-	Datasets     []*Reference      `json:"datasets,omitempty" csv:"datasets" yaml:"datasets,omitempty"`
-	Functions    []*Reference      `json:"functions,omitempty" csv:"functions" yaml:"functions,omitempty"`
-	Models       []*Reference      `json:"models,omitempty" csv:"models" yaml:"models,omitempty"`
-	Dependencies []string          `json:"dependencies,omitempty" csv:"dependencies" yaml:"dependencies,omitempty"`
-	Secrets      Secrets           `json:"secrets,omitempty" csv:"secrets" yaml:"secrets,omitempty"`
+	Version      string               `json:"version,omitempty" csv:"version" yaml:"version,omitempty"`
+	Kind         string               `json:"kind,omitempty" csv:"kind" yaml:"kind,omitempty"`
+	Name         string               `json:"name,omitempty" csv:"name" yaml:"name,omitempty"`
+	Params       map[string]string    `json:"params,omitempty" yaml:"params,omitempty" mapstructure:"params,omitempty"`
+	Metadata     map[string]string    `json:"metadata,omitempty" csv:"metadata" yaml:"metadata,omitempty"`
+	Datasets     []DatasetOrReference `json:"datasets,omitempty" csv:"datasets" yaml:"datasets,omitempty"`
+	Functions    []*Reference         `json:"functions,omitempty" csv:"functions" yaml:"functions,omitempty"`
+	Models       []*Reference         `json:"models,omitempty" csv:"models" yaml:"models,omitempty"`
+	Dependencies []string             `json:"dependencies,omitempty" csv:"dependencies" yaml:"dependencies,omitempty"`
+	Secrets      Secrets              `json:"secrets,omitempty" csv:"secrets" yaml:"secrets,omitempty"`
+}
+
+type DatasetOrReference struct {
+	Dataset   DatasetSpec
```

Comment:
This could be a generic struct which handles the Models too.

Reply:
I think if we largely agree this is the way forward for reading the YAML in Go, I can probably refactor to give it similar naming to Rust (e.g. calling it `Component` instead of `DatasetOrReference`)

Reply:
@y-f-u I've refactored to rename to `Component` and moved it out to another file to make it easier to add more components. I've not added `Model` for now though, as I don't have enough knowledge to know which properties to use to make the unmarshal detect a `Model` vs a `Dataset`.